extensions/attributegen/plugin.h

```diff
@@ -0,0 +1,205 @@
+    if (!evaluateExpression(condition_token_, &matched)) {
+      LOG_WARN(absl::StrCat("Failed to evaluate expression: ", condition_));
+      return {};
+    }
+    if (matched) {
+      LOG_DEBUG(absl::StrCat("Matched: ", condition_));
+    }
+    return matched;
+  };
+
+  std::string value() const { return value_; };
+
+  void cleanup() {
+    if (!condition_.empty()) {
+      exprDelete(condition_token_);
+    }
+  }
+
+ protected:
+  // Only used for debug messages after construction.
+  std::string condition_;
+  // Expression token associated with the condition.
+  uint32_t condition_token_;
+  std::string value_;
+};
+
+enum EvalPhase { on_log, on_request };
+
+class AttributeGenerator {
+ public:
+  explicit AttributeGenerator(EvalPhase phase, std::string output_attribute)
+      : phase_(phase),
+        output_attribute_(output_attribute),
+        error_attribute_(absl::StrCat(output_attribute, "_error")) {}
+
+  // If evaluation is successful returns true and sets result.
+  Optional<bool> evaluate(std::string* val) const {
+    for (const auto& match : matches_) {
+      auto eval_status = match.evaluate();
+      if (!eval_status) {
+        return {};
+      }
+      if (eval_status.value()) {
+        *val = match.value();
+        return true;
+      }
+    }
+    return false;
+  }
+
+  EvalPhase phase() const { return phase_; }
+  std::string output_attribute() const { return output_attribute_; }
+  std::string error_attribute() const { return error_attribute_; }
+  void add_match(Match match) { matches_.push_back(match); }
+  void cleanup() {
+    for (auto match : matches_) {
+      match.cleanup();
+    }
+    matches_.clear();
+  }
+
+ protected:
+  EvalPhase phase_;
+  std::string output_attribute_;
+  std::string error_attribute_;
+  std::vector<Match> matches_;
+};
+
+// PluginRootContext is the root context for all streams processed by the
+// thread. It has the same lifetime as the worker thread and acts as target
+// for interactions that outlives individual stream, e.g. timer, async calls.
+class PluginRootContext : public RootContext {
+ public:
+  PluginRootContext(uint32_t id, StringView root_id)
+      : RootContext(id, root_id) {}
+  ~PluginRootContext() = default;
+
+  bool onConfigure(size_t) override;
+  // is called even if
+  bool onDone() override {
+    cleanupAttributeGen();
+    return true;
+  };
+  void attributeGen(EvalPhase);
+
+ protected:
+  // Destroy host resources for the allocated expressions.
+  void cleanupAttributeGen();
+
+  bool initAttributeGen(const istio::attributegen::PluginConfig& config);
+
+ private:
+  std::vector<AttributeGenerator> gen_;
+};
+
+// Per-stream context.
+class PluginContext : public Context {
+ public:
+  explicit PluginContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  void onLog() override { rootContext()->attributeGen(on_log); };
+
+  FilterHeadersStatus onRequestHeaders(uint32_t) override {
+    rootContext()->attributeGen(on_request);
```

Comment:
There's too much code in the header.  Please move some code into .cc file. That will make imports smaller, too.

Reply:
moved to .cc